src/treelearner/gpu_linear_tree_learner.cpp

```diff
@@ -0,0 +1,381 @@
+/*!
+ * Copyright (c) 2024 Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See LICENSE file in the project root for license information.
+ */
+#include "gpu_linear_tree_learner.h"
+
+#include <Eigen/Dense>
+
+#include <algorithm>
+
+namespace LightGBM {
+
+void GPULinearTreeLearner::Init(const Dataset* train_data, bool is_constant_hessian) {
```

Comment:
Overall, the solution looks good to me. I've read half of the implementation and need some more time to check the correctness. Thank you.

For test case, I think the test case for this scenario is automatically enabled with existing test cases for linear trees and `device_type=gpu`.